File: 프로그래머스/lv4/64063. 호텔 방 배정/호텔 방 배정.py
import sys
sys.setrecursionlimit(10**6)

def find_room(num, hotel):
    if not num in hotel:
        hotel[num] = num + 1
        return num
    else:
        hotel[num] = find_room(hotel[num], hotel)
        return hotel[num]

def solution(k, room_number):
    hotel = {}
    result = []
    
    for num in room_number:
        result.append(find_room(num, hotel))
        
    return result
            

File: 프로그래머스/4/64063. 호텔 방 배정/호텔 방 배정.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

class Solution {
    static HashMap<Long, Long> roomStatus;
    static List<Long> result;

    public long[] solution(long k, long[] room_number) {
        roomStatus = new HashMap<>();
        result = new ArrayList<>();

        for (long room : room_number) {
            findRoom(room);
        }

        return result.stream().mapToLong(i -> i).toArray();
    }

    long findRoom(long room) {
        if (roomStatus.get(room) == null) {
            roomStatus.put(room, room + 1);
            result.add(room);
            return room + 1;
        } else {
            long enableRoom = findRoom(roomStatus.get(room));
            roomStatus.put(room, enableRoom);
            return enableRoom;
        }
    }
}